File: onlyskills-repo/git_to_zkerdfa_monster.pl
% Git Repo to zkerdfa Monster Form
% Maps repo to 196,883-dimensional Monster representation (trimmed to ring)

:- module(git_to_zkerdfa_monster, [
    repo_to_monster_form/2,
    monster_coords/2,
    trim_to_ring/3
]).

% Monster group representation dimension
monster_dim(196883).  % Smallest faithful representation

% Convert git repo to zkerdfa Monster form
repo_to_monster_form(RepoPath, MonsterForm) :-
    % 1. Extract git attributes
    git_attributes(RepoPath, Attrs),
    
    % 2. Map to Monster coordinates (196,883 dims)
    attrs_to_monster_coords(Attrs, FullCoords),
    
    % 3. Trim to ring (71-dimensional subspace)
    trim_to_ring(FullCoords, 71, RingCoords),
    
    % 4. Generate zkerdfa URL
    generate_zkerdfa_url(RingCoords, URL),
    
    % 5. Build Monster form
    MonsterForm = monster_form(
        repo(RepoPath),
        full_dim(196883),
        ring_dim(71),
        coords(RingCoords),
        url(URL)
    ).

% Extract git attributes
git_attributes(RepoPath, Attrs) :-
    % Commits
    git_commit_count(RepoPath, Commits),
    
    % Files
    git_file_count(RepoPath, Files),
    
    % Authors
    git_author_count(RepoPath, Authors),
    
    % Languages
    git_languages(RepoPath, Languages),
    
    % Age (days)
    git_age_days(RepoPath, Age),
    
    % Size (bytes)
    git_size(RepoPath, Size),
    
    Attrs = [
        commits(Commits),
        files(Files),
        authors(Authors),
        languages(Languages),
        age(Age),
        size(Size)
    ].

% Map attributes to Monster coordinates (196,883 dimensions)
attrs_to_monster_coords(Attrs, Coords) :-
    monster_dim(Dim),
    
    % Initialize zero vector
    length(Coords, Dim),
    maplist(=(0), Coords),
    
    % Map each attribute to dimensions
    foldl(map_attr_to_dims, Attrs, Coords, FinalCoords),
    
    Coords = FinalCoords.

% Map single attribute to dimensions
map_attr_to_dims(commits(N), CoordsIn, CoordsOut) :-
    % Commits → first 15 dimensions (Monster primes)
    monster_primes(Primes),
    maplist(commit_to_coord(N), Primes, Values),
    replace_coords(0, Values, CoordsIn, CoordsOut).

map_attr_to_dims(files(N), CoordsIn, CoordsOut) :-
    % Files → dimensions 15-30
    Start is 15,
    file_coords(N, Values),
    replace_coords(Start, Values, CoordsIn, CoordsOut).

map_attr_to_dims(authors(N), CoordsIn, CoordsOut) :-
    % Authors → dimensions 30-45
    Start is 30,
    author_coords(N, Values),
    replace_coords(Start, Values, CoordsIn, CoordsOut).

map_attr_to_dims(languages(Langs), CoordsIn, CoordsOut) :-
    % Languages → dimensions 45-116 (71 language slots)
    Start is 45,
    language_coords(Langs, Values),
    replace_coords(Start, Values, CoordsIn, CoordsOut).

map_attr_to_dims(age(Days), CoordsIn, CoordsOut) :-
    % Age → dimensions 116-131
    Start is 116,
    age_coords(Days, Values),
    replace_coords(Start, Values, CoordsIn, CoordsOut).

map_attr_to_dims(size(Bytes), CoordsIn, CoordsOut) :-
    % Size → dimensions 131-146
    Start is 131,
    size_coords(Bytes, Values),
    replace_coords(Start, Values, CoordsIn, CoordsOut).

% Monster primes (15)
monster_primes([2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 41, 47, 59, 71]).

% Commit to coordinate
commit_to_coord(N, Prime, Value) :-
    Value is N mod Prime.

% File coordinates
file_coords(N, Values) :-
    length(Values, 15),
    maplist(file_coord(N), Values).

file_coord(N, Value) :-
    Value is N mod 71.

% Author coordinates
author_coords(N, Values) :-
    length(Values, 15),
    maplist(author_coord(N), Values).

author_coord(N, Value) :-
    Value is (N * 2) mod 71.

% Language coordinates (71 slots for languages)
language_coords(Langs, Values) :-
    length(Values, 71),
    maplist(=(0), Values),
    % Set 1 for each language present
    foldl(set_language, Langs, Values, FinalValues),
    Values = FinalValues.

set_language(Lang, ValsIn, ValsOut) :-
    language_index(Lang, Idx),
    replace_nth(Idx, 1, ValsIn, ValsOut).

% Language to index mapping
language_index(rust, 71).
language_index(lean, 59).
language_index(prolog, 47).
language_index(nix, 41).
language_index(python, 2).
language_index(_, 11).  % Default

% Age coordinates
age_coords(Days, Values) :-
    length(Values, 15),
    maplist(age_coord(Days), Values).

age_coord(Days, Value) :-
    Value is (Days // 365) mod 71.

% Size coordinates
size_coords(Bytes, Values) :-
    length(Values, 15),
    maplist(size_coord(Bytes), Values).

size_coord(Bytes, Value) :-
    MB is Bytes // (1024 * 1024),
    Value is MB mod 71.

% Trim to ring (project to 71-dimensional subspace)
trim_to_ring(FullCoords, RingDim, RingCoords) :-
    % Take first RingDim coordinates
    length(RingCoords, RingDim),
    append(RingCoords, _, FullCoords).

% Replace coordinates
replace_coords(Start, Values, CoordsIn, CoordsOut) :-
    length(Before, Start),
    append(Before, Rest, CoordsIn),
    length(Values, Len),
    length(ToReplace, Len),
    append(ToReplace, After, Rest),
    append(Values, After, NewRest),
    append(Before, NewRest, CoordsOut).

% Replace nth element
replace_nth(0, Val, [_|T], [Val|T]) :- !.
replace_nth(N, Val, [H|T], [H|R]) :-
    N > 0,
    N1 is N - 1,
    replace_nth(N1, Val, T, R).

% Generate zkerdfa URL
generate_zkerdfa_url(Coords, URL) :-
    % Encode coordinates as base64
    coords_to_bytes(Coords, Bytes),
    base64_encode(Bytes, Encoded),
    
    % Hash for short identifier
    hash_coords(Coords, Hash),
    
    format(atom(URL),
        'zkerdfa://monster/ring71/~w?coords=~w',
        [Hash, Encoded]).

% Coords to bytes
coords_to_bytes(Coords, Bytes) :-
    maplist(coord_to_byte, Coords, Bytes).

coord_to_byte(Coord, Byte) :-
    Byte is Coord mod 256.

% Hash coordinates
hash_coords(Coords, Hash) :-
    sumlist(Coords, Sum),
    Hash is Sum mod (71 * 71 * 71).

% Base64 encode (simplified)
base64_encode(Bytes, Encoded) :-
    atom_codes(Encoded, Bytes).

% Git operations (stubs - implement with shell calls)
git_commit_count(_, 100).
git_file_count(_, 50).
git_author_count(_, 5).
git_languages(_, [rust, prolog, lean]).
git_age_days(_, 365).
git_size(_, 10485760).  % 10 MB

% Example usage:
% ?- repo_to_monster_form('/path/to/repo', Form).
% Form = monster_form(repo(...), full_dim(196883), ring_dim(71), coords([...]), url(...))

File: onlyskills-repo/src/bin/git_to_zkerdfa_monster.rs
// Git Repo to zkerdfa Monster Form (Rust)
use std::path::Path;
use std::process::Command;
use serde::{Deserialize, Serialize};

const MONSTER_DIM: usize = 196883;  // Smallest faithful representation
const RING_DIM: usize = 71;         // Trimmed to Monster prime ring

const MONSTER_PRIMES: [u8; 15] = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 41, 47, 59, 71];

#[derive(Debug, Serialize, Deserialize)]
pub struct MonsterForm {
    pub repo_path: String,
    pub full_dim: usize,
    pub ring_dim: usize,
    pub coords: Vec<u8>,
    pub url: String,
}

#[derive(Debug)]
struct GitAttributes {
    commits: u64,
    files: u64,
    authors: u64,
    languages: Vec<String>,
    age_days: u64,
    size_bytes: u64,
}

pub fn repo_to_monster_form(repo_path: &Path) -> MonsterForm {
    // 1. Extract git attributes
    let attrs = extract_git_attributes(repo_path);
    
    // 2. Map to Monster coordinates (196,883 dims)
    let full_coords = attrs_to_monster_coords(&attrs);
    
    // 3. Trim to ring (71 dims)
    let ring_coords = trim_to_ring(&full_coords, RING_DIM);
    
    // 4. Generate zkerdfa URL
    let url = generate_zkerdfa_url(&ring_coords);
    
    MonsterForm {
        repo_path: repo_path.display().to_string(),
        full_dim: MONSTER_DIM,
        ring_dim: RING_DIM,
        coords: ring_coords,
        url,
    }
}

fn extract_git_attributes(repo_path: &Path) -> GitAttributes {
    GitAttributes {
        commits: git_commit_count(repo_path),
        files: git_file_count(repo_path),
        authors: git_author_count(repo_path),
        languages: git_languages(repo_path),
        age_days: git_age_days(repo_path),
        size_bytes: git_size(repo_path),
    }
}

fn attrs_to_monster_coords(attrs: &GitAttributes) -> Vec<u8> {
    let mut coords = vec![0u8; MONSTER_DIM];
    
    // Commits → dimensions 0-14 (Monster primes)
    for (i, &prime) in MONSTER_PRIMES.iter().enumerate() {
        coords[i] = (attrs.commits % prime as u64) as u8;
    }
    
    // Files → dimensions 15-29
    for i in 0..15 {
        coords[15 + i] = (attrs.files % 71) as u8;
    }
    
    // Authors → dimensions 30-44
    for i in 0..15 {
        coords[30 + i] = ((attrs.authors * 2) % 71) as u8;
    }
    
    // Languages → dimensions 45-115 (71 language slots)
    for lang in &attrs.languages {
        let idx = language_index(lang);
        if idx < 71 {
            coords[45 + idx] = 1;
        }
    }
    
    // Age → dimensions 116-130
    let years = attrs.age_days / 365;
    for i in 0..15 {
        coords[116 + i] = (years % 71) as u8;
    }
    
    // Size → dimensions 131-145
    let mb = attrs.size_bytes / (1024 * 1024);
    for i in 0..15 {
        coords[131 + i] = (mb % 71) as u8;
    }
    
    coords
}

fn trim_to_ring(full_coords: &[u8], ring_dim: usize) -> Vec<u8> {
    // Project to ring: take first ring_dim coordinates
    full_coords[..ring_dim].to_vec()
}

fn generate_zkerdfa_url(coords: &[u8]) -> String {
    // Hash coordinates
    let hash: u64 = coords.iter().map(|&c| c as u64).sum::<u64>() % (71 * 71 * 71);
    
    // Encode coordinates as base64
    let encoded = base64::encode(coords);
    
    format!("zkerdfa://monster/ring71/{}?coords={}", hash, encoded)
}

fn language_index(lang: &str) -> usize {
    match lang {
        "rust" => 70,
        "lean" => 58,
        "prolog" => 46,
        "nix" => 40,
        "python" => 1,
        _ => 10,
    }
}

// Git operations
fn git_commit_count(repo_path: &Path) -> u64 {
    let output = Command::new("git")
        .args(&["-C", repo_path.to_str().unwrap(), "rev-list", "--count", "HEAD"])
        .output()
        .ok();
    
    output
        .and_then(|o| String::from_utf8(o.stdout).ok())
        .and_then(|s| s.trim().parse().ok())
        .unwrap_or(0)
}

fn git_file_count(repo_path: &Path) -> u64 {
    let output = Command::new("git")
        .args(&["-C", repo_path.to_str().unwrap(), "ls-files"])
        .output()
        .ok();
    
    output
        .map(|o| o.stdout.iter().filter(|&&b| b == b'\n').count() as u64)
        .unwrap_or(0)
}

fn git_author_count(repo_path: &Path) -> u64 {
    let output = Command::new("git")
        .args(&["-C", repo_path.to_str().unwrap(), "shortlog", "-sn", "HEAD"])
        .output()
        .ok();
    
    output
        .map(|o| o.stdout.iter().filter(|&&b| b == b'\n').count() as u64)
        .unwrap_or(0)
}

fn git_languages(repo_path: &Path) -> Vec<String> {
    // Simplified: detect by file extensions
    vec!["rust".to_string(), "prolog".to_string()]
}

fn git_age_days(repo_path: &Path) -> u64 {
    // First commit timestamp
    365 // Placeholder
}

fn git_size(repo_path: &Path) -> u64 {
    // .git directory size
    10 * 1024 * 1024 // Placeholder
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_monster_form() {
        let path = Path::new(".");
        let form = repo_to_monster_form(path);
        
        assert_eq!(form.full_dim, 196883);
        assert_eq!(form.ring_dim, 71);
        assert_eq!(form.coords.len(), 71);
        assert!(form.url.starts_with("zkerdfa://monster/ring71/"));
    }
    
    #[test]
    fn test_trim_to_ring() {
        let full = vec![1u8; 196883];
        let ring = trim_to_ring(&full, 71);
        
        assert_eq!(ring.len(), 71);
        assert_eq!(ring[0], 1);
    }
}
